### ISSUE TITLE
Push Notifications entitlement getting added automatically in prebuild.

### ISSUE THREAD

**Issue by sainjay** [expo/expo#27668]
### Minimal reproducible example

https://github.com/sainjay/Expo-app-50

### What platform(s) does this occur on?

iOS

### Did you reproduce this issue in a development build?

Yes

### Summary

App.json doesn't specify anything for entitlements still the `aps-environment` is getting added when prebuild is ran.

Steps to reproduce:
1. npx create-expo-app my-app-50
2. cd my-app-50
3. npx expo prebuild --platform ios --clean
4. cat ios/myapp/myapp.entitlements

<img width="785" alt="image" src="https://github.com/expo/expo/assets/136019884/873a5a67-011e-48f7-ba01-64a56a348f89">

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
      Yarn: 3.6.1 - ~/.nvm/versions/node/v18.16.0/bin/yarn
      npm: 9.5.1 - ~/.nvm/versions/node/v18.16.0/bin/npm
      Watchman: 2023.10.23.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.13.0 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, visionOS 1.0, watchOS 10.2
      Android SDK:
        API Levels: 28, 30, 31, 32, 33, 34
        Build Tools: 30.0.3, 33.0.0, 33.0.1, 34.0.0
        System Images: android-30 | Google Play ARM 64 v8a, android-31 | Google APIs ARM 64 v8a, android-34 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2023.2 AI-232.10227.8.2321.11479570
      Xcode: 15.2/15C500b - /usr/bin/xcodebuild
    npmPackages:
      expo: ~50.0.13 => 50.0.13 
      react: 18.2.0 => 18.2.0 
      react-native: 0.73.5 => 0.73.5 
    Expo Workflow: managed

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
(node:25679) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node --trace-warnings ...` to show where the warning was created)
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check native tooling versions
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!

**Comment by douglowder** [comment]
That entitlement is set by default in prebuild: https://github.com/expo/expo/blob/sdk-50/packages/%40expo/prebuild-config/src/plugins/unversioned/expo-notifications/expo-notifications.ts . This is not new behavior, this has been in the code for several years now.

I'll discuss this internally with the team to see if this needs to be adjusted. It would be helpful to know what issues or problems arise for you when this entitlement is set.

**Comment by sainjay** [comment]
Hi @douglowder, 
Only issue I faced was that `Push Notification` entitlement had to be added to Mobile Provisioning even though I'm not using Push Notifications currently. I tried passing empty array for entitlement in app.json but it didn't change anything when Provisioning didn't have the entitlement. 
Any clue for why this entitlement `aps-environment` is added always in prebuild?

**Comment by douglowder** [comment]
@sainjay sorry for the delay... we took some time to go over the reasons why the prebuild code was added several years ago, and have determined that it is definitely no longer needed. I have put up a PR to remove it.

**Comment by sainjay** [comment]
Hi @douglowder , Thanks for the quick update and the new PR. Currently i updated my Mobile Provisioning to have `Push Notifications` enabled from Apple Developer Console. Once the PR gets merged, how do we manually add the Entitlement in app.json. Is following the correct format?

"ios": {
   "entitlements": ["aps-environment"]
}

Kindly confirm the above format or provide the update for app.json. Might help other developers as well.

**Comment by douglowder** [comment]
The `entitlements` property is an object, not an array. In your case, you want to set `aps-environment` manually, so your `app.json` should look like this:

```json
    "ios": {
      "entitlements": {
        "aps-environment": "development"
      },
    },
```

If later, you decide to implement push notifications in your app, and you add the `expo-notifications` package to your project, then this property will be set automatically by the plugin in that package, and you won't need a custom `entitlements` object.

Documentation on the `entitlements` format: https://docs.expo.dev/versions/latest/config/app/#entitlements

**Comment by douglowder** [comment]
If you are using EAS to build your app, this documentation may also be helpful: https://docs.expo.dev/build-reference/ios-capabilities/

**Comment by sainjay** [comment]
Thanks @douglowder. This is helpful. Also facing issues with Expo-Location along with many others. Can you ask the proper team members to take a look at this issue from 2020 which still exists:

https://github.com/expo/expo/issues/10756

Proper blog or video on youtube for implementing Expo-Location may help. Currently people are facing delay in fetching location using `getCurrentPositionAsync` for upto 15secs or more sometimes which is too slow.

**Comment by younes0** [comment]
@douglowder would it be possible to mention this breaking change on the SDK 51 changelog ? 
I think it's an important one as it can break notifications for projects using a library like `react-native-firebase` (actually it did for us 🥲)

**Comment by sainjay** [comment]
@younes0  you can add Push Notification Entitlement back with this simple change in `app.json`.

```
"ios": {
      "entitlements": {
        "aps-environment": "development"
      },
    },
```

**Comment by brentvatne** [comment]
@younes0 - sorry for the trouble, i've pushed a change to add this to our release notes, it'll be deployed shortly

**Comment by git-commit** [comment]
On our expo project we get a warning from app store connect that the entitlement is missing. Is anybody else getting this? We're not using push notifications in our app or dependencies.

**Comment by douglowder** [comment]
> In our expo project we get a warning from app store connect (ITMS-90078) that the entitlement is missing. Is anybody else getting this? We're not using push notifications in our app or dependencies.

@git-commit Please see https://github.com/expo/fyi/blob/main/apns-entitlement-sdk-51.md for a detailed explanation. Also, kudos for a most excellent GitHub handle 😄

**Comment by gkasireddy202** [comment]
I am getting a warning message from App Store Connect.I did not use the push notification in my app.
ITMS-90078: Missing Push Notification Entitlement - Your app appears to register with the Apple Push Notification service, but the app signature's entitlements do not include the 'aps-environment' entitlement. If your app uses the Apple Push Notification service, make sure your App ID is enabled for Push Notification in the Provisioning Portal, and resubmit after signing your app with a Distribution provisioning profile that includes the 'aps-environment' entitlement. Xcode does not automatically copy the aps-environment entitlement from provisioning profiles at build time. This behavior is intentional. To use this entitlement, either enable Push Notifications in the project editor's Capabilities pane, or manually add the entitlement to your entitlements file. For more information, see

react-native:0.73.7
iOS:17
expo: "50.0.17",
expo-local-authentication: "13.8.0",

**Comment by Albert-Einstein-Nine** [comment]
Same as @gkasireddy202. I got the same email from Apple when I distributed a new build. 
I tried asking for no permissions in the `app.json` just as below, but that didn't resolve the issue. Any idea how I can resolve this issue with Apple?

>     "ios": {
>       "permissions": {
>         "notifications": false
>       }

**Comment by NicolasRecomiendApp** [comment]
Hi everyone! I'm using react-native-firebase/messaging, so I'm getting a build error if i set my app.json as: 

```
"ios": {
      "entitlements": {
        "aps-enviroment": "development"
      },
      ...
      }
```
If I don't add it, the  await messaging().getToken(); retrieves me a warning: [Error: [messaging/unknown] missing  valid authorisation key ‘aps-environment’ for the application].

It was all correctly configured before we upgraded to SDK51 (we have a production running well on 50).

Are we missing something? Thanks in advance!

**Comment by sainjay** [comment]
I noticed that you're missing `n` in `aps-environment`. Not sure if that is the build error that you're facing, but you can try the below. I'm facing a separate build error with Firebase for `chrono file not found`

```
"ios": {
      "entitlements": {
        "aps-environment": "development"
      },
      ...
      }
 ```

**Comment by NicolasRecomiendApp** [comment]
That was the problem, thanks!! @sainjay

**Comment by uloco** [comment]
I am on sdk 52 and this is still happening to me...

**Comment by Mykyta-Chernenko** [comment]
I bumped into this thread again because I couldn't remember how I resolved this one a couple of months before

here is how it should be fixed, the new versions of expo-notifications add the needed config.

app.json

```
plugins: [
....
      ["expo-notifications"],
...
]
```

**Comment by thorgas** [comment]
For anyone hitting this with a personal Apple development team: expo-notifications adds aps-environment unconditionally, and a personal team cannot provision the Push Notifications capability, so prebuild breaks local signing even for apps that only schedule local notifications.

I published a small config plugin that removes the entitlement: [expo-no-push-entitlement](https://www.npmjs.com/package/expo-no-push-entitlement). Tbh. Claude implemented it, but I'm using it in my own app and its working fine.

One thing worth noting: it must be registered before expo-notifications in plugins.
Not a substitute for a first-party opt-out, but it unblocks local-notification-only apps today.
